### DISCUSSION TITLE
How does egui painting system really work?

### DISCUSSION THREAD

**Discussion by Henrique194** [emilk/egui#3852] [Q&A]
Take for instance the following code example:
```
struct MyApp;

impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::CentralPanel::default().show(ctx, |ui| {
            ui.label("Some text");
        });
    }
}
```

How does egui know where to place the label?
If I have a panel of size 500x500, do I have control to put the label at some specific location?
Pixel level alteration is only possible through the renderer system (glow/wgpu) ?

**Comment by YgorSouza** [comment]
Every widget's implementation has to call one of the Ui::allocate_* methods to allocate space for itself, then check for input in that allocated space and change its state accordingly, then use the UI painter to draw what it needs to display. There is a simple example in the [demo app](https://github.com/emilk/egui/blob/master/crates/egui_demo_lib/src/demo/toggle_switch.rs). Every time you allocate space, the cursor moves according to the current layout (vertical, horizontal, wrapped, justified etc), so the next time you allocate, you will get a rect adjacent to the previous one in the correct direction and with the correct spacing. If you want to ignore the layout and just allocate some specific area of the Ui, you can use [allocate_rect](https://docs.rs/egui/0.25.0/egui/struct.Ui.html#method.allocate_rect). There are a few other convenience methods listed on that page.

**Comment by Henrique194** [comment]
That clears things up, thanks. The painter then must call the renderer right? So if I want to change some random pixel in the screen, I should call the renderer context ?

**Reply by YgorSouza** [reply]
I'm not so sure about that part. There's actually already a question about it here, but it has no answer. egui doesn't render anything by itself, it outputs a [FullOutput](https://docs.rs/egui/0.25.0/egui/output/struct.FullOutput.html) struct after each frame, which tells the backend what to draw. It contains a list of shapes and a list of textures. So I guess if you wanted to change an arbitrary pixel in a backend-agnostic way, it would have to be either a shape (i.e., a small `Rect`) or a texture (i.e., an image created from an RGBA array, in the manner explained in #3426). Maybe someone else knows better, though.

**Reply by YgorSouza** [reply]
Or if you're only using egui with a specific backend (like eframe does), you could call the backend directly. I guess an example of that would be https://www.egui.rs/#Custom3d

**Reply by Henrique194** [reply]
Thanks for the answers, I will give a try using the backend directly.
